src/translator/annotation.h

```diff
@@ -17,83 +17,99 @@ struct ByteRange {
-/// is expected to be read only in this mode of operation.
+/// Since we are using SentencePiece, a token can include whitespace.  The term
+/// "word" is used, somewhat incorrectly, as a synonym of token.
 ///
-/// **Idea**
-///
-/// Annotation is intended to be the same structure conceptually as below,
-/// except the `std::vector<std::vector<ByteRange>>` hammered into a flat
-/// structure to avoid multiple reallocs keeping efficiency in mind. This is
-/// achieved by having markers of where sentence ends in the flat container
-/// storing word ByteRanges.
-///
-/// ```cpp
-/// typedef ByteRange Word;
-/// // std::vector<ByteRange>, a single sentence
-/// typedef std::vector<Word> Sentence;
-/// std::vector<std::vector<ByteRange> // multiple sentences
-/// typedef std::vector<Sentence> Annotation;
-///
-/// Annotation example;
-/// ```
-/// This structure exists to provide a consistent API to access the nested
-/// sentences of varying lengths, which occur in source-text processed into
-/// multiple sentences, and target-text translated from source as multiple
-/// sentences, both composed of (sub)-words, providing a List[List] like access
-/// while storing it in a compact and efficient manner.
+/// A gap can be empty (for example there may not have been whitespace at the
+/// beginning).  A sentence can also be empty (typically the translation system
+/// produced empty output).  That's fine, these are just empty ranges as you
+/// would expect.
 class Annotation {
 public:
-  /// Annotation is constructed empty. See `addSentence()` to populate it with
-  /// annotations.
+  /// Initially an empty string.  Populated by AnnotatedText.
   Annotation() {
-    // The -1-th sentence ends at 0.
-    sentenceEndIds_.push_back(0);
+    token_begin_.push_back(0);
+    token_begin_.push_back(0);
+    gap_.push_back(0);
   }
 
-  size_t numSentences() const { return sentenceEndIds_.size() - 1; }
+  size_t numSentences() const { return gap_.size() - 1; }
 
   /// Returns number of words in the sentence identified by `sentenceIdx`.
-  size_t numWords(size_t sentenceIdx) const;
-
-  /// Adds a sentences from `vector<ByteRange>` representation, internally doing
-  /// extra book-keeping for the sentence terminal markings. Sentences are
-  /// expected to be added in order as they occur in text.
-  void addSentence(std::vector<ByteRange> &sentence);
+  size_t numWords(size_t sentenceIdx) const {
+    return gap_[sentenceIdx + 1] - gap_[sentenceIdx] - 1 /* minus the gap */;
+  }
 
   /// Returns a ByteRange representing `wordIdx` in sentence indexed by
   /// `sentenceIdx`. `wordIdx` follows 0-based indexing, and should be less than
   /// `.numWords()` for `sentenceIdx` for defined behaviour.
-  ByteRange word(size_t sentenceIdx, size_t wordIdx) const;
+  ByteRange word(size_t sentenceIdx, size_t wordIdx) const {
+    size_t tokenIdx = gap_[sentenceIdx] + 1 + wordIdx;
+    return ByteRange {token_begin_[tokenIdx], token_begin_[tokenIdx + 1]};
```

Comment:
Since we're doing asserts, there's an assert missing here described loosely as:

```
gap_[sentenceIdx] + 1 + wordIdx < gap_[sentenceIdx+1] + 1
```